### PR TITLE
Decodes URL Safe Parameters

### PR DESCRIPTION
**Pull request by scottyplunkett**
- Adds decoding to query values for request w/ params
- Outputs correctly to dynamic HTML
- Changed Pages to reflect expected variables in CobSpec test

### CODE REVIEW

File: pages/dynamic.html [hunk 1]
```diff
@@ -6,8 +6,8 @@
 <body>
 <h1>
     Hello,
-    <i>$first</i>
-    <i>$last</i>!
+    <i>$variable_1</i>
```
Comment (NicoleCarpenter) [inline pages/dynamic.html:9]: Just curious, why did you change this? For the test?

File: test/com/scottyplunkett/server/ParserTest.java [hunk 2]
```diff
@@ -55,4 +56,10 @@ void parseForMethod() {
         assertEquals("DELETE", Parser.parseFor("method", deleteRequest));
     }
 
+    @Test
+    void decodeURLSafeString() throws UnsupportedEncodingException {
```
Comment (NicoleCarpenter) [inline test/com/scottyplunkett/server/ParserTest.java:60]: I'm ok with this

**Review by NicoleCarpenter — APPROVED** [review]

File: pages/dynamic.html (hunk 1 above)
Comment (esmevane) [reply pages/dynamic.html:9]: I was wondering the same thing.

File: src/com/scottyplunkett/server/Parser.java [hunk 3]
```diff
@@ -18,16 +20,22 @@ static String findRequestMethod(String requestLine) {
...
 
-    static Map<String, String> mapQuery(String requestLine) {
+    static Map<String, String> mapQuery(String requestLine) throws UnsupportedEncodingException {
         Map<String, String> query = new HashMap<>();
         String justQuery = splitByQuestionMark.apply(splitBySpace.apply(requestLine)[1])[1];
         for (String keyValue : justQuery.split("\\&")){
             String[] pairs = keyValue.split("\\=");
-            query.put(pairs[0], pairs.length == 1 ? "" : pairs[1]);
+            String decodedParam = decodeURLSafeString(pairs[1]);
+            query.put(pairs[0], pairs.length == 1 ? "" : decodedParam);
```
Comment (esmevane) [inline src/com/scottyplunkett/server/Parser.java:33]: Arrays (and fetching info from arrays with index integers) can be a pretty volatile risk vector in a lot of projects.  This is probably fine here, but I feel like there's an opportunity to maybe shape this process a little more as we go.

**Review by esmevane — APPROVED** [review]
This was a great PR 🤘 Compact, easy to digest, clear code, tests.  Awesome stuff!

One thing I like here is that you're strengthening your boundaries.

For example, Parser is a formal boundary, where you're introducing a layer that converts user input into domain objects.  I think one way you could possibly strengthen it further is by taking these strings and maps that it's generating, and sort of coercing them into _actual Value objects_.

It's not like, 100% of the time, but often enough I really come to regret keeping things primitives.  The strings turn into things I want to know about or turn into other things, and the index fetching starts to bleed across other objects, etc.

[Here's a nice wiki hole you can fall into about it, where way smarter people than me have chimed in.](http://wiki.c2.com/?PrimitiveObsession)
